src/node_trace_events.cc

```diff
@@ -165,9 +169,97 @@ void NodeCategorySet::RegisterExternalReferences(
+            u8)
+      .Check();
+}
+
+bool BindingData::PrepareForSerialization(v8::Local<v8::Context> context,
+                                          v8::SnapshotCreator* creator) {
+  return true;
+}
+
+InternalFieldInfoBase* BindingData::Serialize(int index) {
+  DCHECK_EQ(index, BaseObject::kEmbedderType);
+  InternalFieldInfo* info =
+      InternalFieldInfoBase::New<InternalFieldInfo>(type());
+  return info;
+}
+
+void BindingData::Deserialize(v8::Local<v8::Context> context,
+                              v8::Local<v8::Object> holder,
+                              int index,
+                              InternalFieldInfoBase* info) {
+  DCHECK_EQ(index, BaseObject::kEmbedderType);
+  v8::HandleScope scope(context->GetIsolate());
+  Realm* realm = Realm::GetCurrent(context);
+  BindingData* binding = realm->AddBindingData<BindingData>(context, holder);
+  CHECK_NOT_NULL(binding);
+}
+
+void BindingData::CreatePerIsolateProperties(IsolateData* isolate_data,
+                                             Local<ObjectTemplate> ctor) {}
+
+void BindingData::CreatePerContextProperties(Local<Object> target,
+                                             Local<Value> unused,
+                                             Local<Context> context,
+                                             void* priv) {
+  Realm* realm = Realm::GetCurrent(context);
+  realm->AddBindingData<BindingData>(context, target);
+}
+
+void BindingData::RegisterExternalReferences(
+    ExternalReferenceRegistry* registry) {}
+
+static void CreatePerIsolateProperties(IsolateData* isolate_data,
+                                       Local<ObjectTemplate> ctor) {
+  BindingData::CreatePerIsolateProperties(isolate_data, ctor);
+}
+
+static void CreatePerContextProperties(Local<Object> target,
+                                       Local<Value> unused,
+                                       Local<Context> context,
+                                       void* priv) {
+  BindingData::CreatePerContextProperties(target, unused, context, priv);
+
+  NodeCategorySet::Initialize(target, unused, context, priv);
+}
+
+void RegisterExternalReferences(ExternalReferenceRegistry* registry) {
+  NodeCategorySet::RegisterExternalReferences(registry);
+  BindingData::RegisterExternalReferences(registry);
+}
+
+}  // namespace trace_events
+
 }  // namespace node
 
-NODE_BINDING_CONTEXT_AWARE_INTERNAL(trace_events,
-                                    node::NodeCategorySet::Initialize)
-NODE_BINDING_EXTERNAL_REFERENCE(
-    trace_events, node::NodeCategorySet::RegisterExternalReferences)
+NODE_BINDING_CONTEXT_AWARE_INTERNAL(
+    trace_events, node::trace_events::CreatePerContextProperties)
+NODE_BINDING_PER_ISOLATE_INIT(trace_events,
+                              node::trace_events::CreatePerIsolateProperties)
+NODE_BINDING_EXTERNAL_REFERENCE(trace_events,
+                                node::trace_events::RegisterExternalReferences)
```

Comment:
We can keep the `NODE_BINDING_EXTERNAL_REFERENCE` part unchanged because `bindingdata::RegisterExternalReferences()` is a no-op, right?

Reply:
When I tried, I could not compile without this method inside `BindingData`, I can try again since you suggested `SET_NO_MEMORY_INFO` but it will probably fail.

Reply:
You could try applying this patch:

<details>

```diff
diff --git a/src/node_trace_events.cc b/src/node_trace_events.cc
index d639989016..645dcbed3f 100644
--- a/src/node_trace_events.cc
+++ b/src/node_trace_events.cc
@@ -226,9 +226,6 @@ void BindingData::CreatePerContextProperties(Local<Object> target,
   realm->AddBindingData<BindingData>(context, target);
 }

-void BindingData::RegisterExternalReferences(
-    ExternalReferenceRegistry* registry) {}
-
 static void CreatePerContextProperties(Local<Object> target,
                                        Local<Value> unused,
                                        Local<Context> context,
@@ -238,16 +235,11 @@ static void CreatePerContextProperties(Local<Object> target,
   NodeCategorySet::Initialize(target, unused, context, priv);
 }

-void RegisterExternalReferences(ExternalReferenceRegistry* registry) {
-  NodeCategorySet::RegisterExternalReferences(registry);
-  BindingData::RegisterExternalReferences(registry);
-}
-
 }  // namespace trace_events

 }  // namespace node

 NODE_BINDING_CONTEXT_AWARE_INTERNAL(
     trace_events, node::trace_events::CreatePerContextProperties)
-NODE_BINDING_EXTERNAL_REFERENCE(trace_events,
-                                node::trace_events::RegisterExternalReferences)
+NODE_BINDING_EXTERNAL_REFERENCE(
+    trace_events, node::NodeCategorySet::RegisterExternalReferences)
diff --git a/src/node_trace_events.h b/src/node_trace_events.h
index 4a704bc9f8..22fa420778 100644
--- a/src/node_trace_events.h
+++ b/src/node_trace_events.h
@@ -33,7 +33,6 @@ class BindingData : public SnapshotableObject {
                                          v8::Local<v8::Value> unused,
                                          v8::Local<v8::Context> context,
                                          void* priv);
-  static void RegisterExternalReferences(ExternalReferenceRegistry* registry);
 };

 }  // namespace trace_events
```

</details>

It worked for me locally